plugin/pkg/proxy/connect.go

```diff
@@ -117,6 +118,15 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 	for {
 		ret, err = pc.c.ReadMsg()
 		if err != nil {
+			// If the error is an overflow, we probably have a misbehaving upstream
+			// that is sending out-of-spec (>512 byte) UDP responses without an eDNS0
+			// OPT RR. Set the truncate bit and return the result - this will make the
```

Comment:
Thanks for the PR! This sounds like a simpler solution than having CoreDNS retry TCP itself.

My question is, will this handle the situation where an upstream sets the truncated bit, but actually doesn't fully truncate the result? In my case, the upstream does appear to set the TC bit and partially truncate the response, but it doesn't apply any truncation on the Authority RRs, which pushes it over the bufsize requested (and overflows).

Reply:
It should, but I think it would actually be cleaner to build a new response object that we guarantee is in a clean state. It'll be in the next commit.

I changed to that method based on the documentation:

> ReadMsg reads a message from the connection co. If the received message contains a TSIG record the transaction signature is verified. This method always tries to return the message, **however if an error is returned there are no guarantees that the returned message is a valid representation of the packet read.**